airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/record_buffer/SerializedRecordBufferingStrategy.java

```diff
@@ -0,0 +1,126 @@
+  private final CheckedBiConsumer<AirbyteStreamNameNamespacePair, RecordBufferImplementation, Exception> onStreamFlush;
+  private VoidCallable onFlushEventHook;
+
+  private Map<AirbyteStreamNameNamespacePair, RecordBufferImplementation> allBuffers = new HashMap<>();
+  private long totalBufferSizeInBytes;
+  private final ConfiguredAirbyteCatalog catalog;
+
+  public SerializedRecordBufferingStrategy(final CheckedBiFunction<AirbyteStreamNameNamespacePair, ConfiguredAirbyteCatalog, RecordBufferImplementation, Exception> onCreateBuffer,
+                                           final ConfiguredAirbyteCatalog catalog,
+                                           final CheckedBiConsumer<AirbyteStreamNameNamespacePair, RecordBufferImplementation, Exception> onStreamFlush) {
+    this.onCreateBuffer = onCreateBuffer;
+    this.catalog = catalog;
+    this.onStreamFlush = onStreamFlush;
+    this.totalBufferSizeInBytes = 0;
+    this.onFlushEventHook = null;
+  }
+
+  @Override
+  public void registerFlushEventHook(final VoidCallable onFlushEventHook) {
+    this.onFlushEventHook = onFlushEventHook;
+  }
+
+  @Override
+  public void addRecord(final AirbyteStreamNameNamespacePair stream, final AirbyteMessage message) throws Exception {
+
+    final RecordBufferImplementation streamBuffer = allBuffers.computeIfAbsent(stream, k -> {
+      LOGGER.info("Starting a new buffer for stream {} (current state: {} in {} buffers)",
+          stream.getName(),
+          FileUtils.byteCountToDisplaySize(totalBufferSizeInBytes),
+          allBuffers.size());
+      try {
+        return onCreateBuffer.apply(stream, catalog);
+      } catch (final Exception e) {
+        LOGGER.error("Failed to create a new buffer for stream {}", stream.getName(), e);
+        throw new RuntimeException(e);
+      }
+    });
+    if (streamBuffer != null) {
+      final Long actualMessageSizeInBytes = streamBuffer.accept(message.getRecord());
+      totalBufferSizeInBytes += actualMessageSizeInBytes;
+      if (totalBufferSizeInBytes > streamBuffer.getMaxTotalBufferSizeInBytes()
+          || allBuffers.size() > streamBuffer.getMaxConcurrentStreamsInBuffer()) {
+        flushAll();
+        totalBufferSizeInBytes = 0;
+      } else if (streamBuffer.getCount() > streamBuffer.getMaxPerStreamBufferSizeInBytes()) {
+        flushWriter(stream, streamBuffer);
+      }
+    }
+  }
+
+  @Override
+  public void flushWriter(final AirbyteStreamNameNamespacePair stream, final RecordBufferImplementation writer) throws Exception {
+    LOGGER.info("Flushing buffer of stream {} ({})", stream.getName(), FileUtils.byteCountToDisplaySize(writer.getCount()));
+    onStreamFlush.accept(stream, writer);
+    totalBufferSizeInBytes -= writer.getCount();
+    allBuffers.remove(stream);
+  }
+
+  @Override
+  public void flushAll() throws Exception {
+    LOGGER.info("Flushing all {} current buffers ({} in total)", allBuffers.size(), FileUtils.byteCountToDisplaySize(totalBufferSizeInBytes));
```

Comment:
is there a reason not to call `AirbyteSentry.executeWithTracing("FlushBuffer"` like we do in `DefaultRecordBufferingStrategy`?

Reply:
Good point, I haven't paid attention to the Sentry wrapping calls, let me add them